### PR TITLE
[Fix] Load pack-quantized fc weights for DFlash2 drafts (#39087)

### LINKED ISSUE #39087: [Bug] Quantized DFlash2 draft silently yields ~0% acceptance — no error, no warning (the quiet counterpart to #36599)
### Summary

A **quantized** DFlash2 draft checkpoint loads into `DFlash2DraftModel`, serves, and produces
drafts that are rejected ~100% of the time. There is **no error and no warning** — acceptance
simply collapses from ~3.7 to ~1.0 and decode drops below no-drafter speed.

The same weights **unquantized** work correctly. This is the silent counterpart to #36599 (quantized
NextN draft, `modelopt_fp4`), which fails loudly with a tensor-shape mismatch instead.

This is a read-only report — I have not root-caused it and am not opening a PR.

### Repro

Target: `Frozenlock/Qwen3.8-27B-int4-AutoRound` (auto-round, 4-bit, g128, sym).
Two drafters that are **the same model**, differing only in quantization:

| drafter | params | dtypes | quant |
|---|---:|---|---|
| `z-lab/Qwen3.8-27B-DFlash2` | 1,924,404,480 | BF16 | none |
| `syvai/Qwen3.8-27B-DFlash2-W4A16` | 1,924,404,480 | I32 + BF16 | compressed-tensors |

Identical `architectures` (`DFlash2DraftModel`), identical `dflash_config`
(`block_size 8, selector_rank 256, selector_top_k 16`, same `target_layer_ids`), identical tensor
names including `candidate_selector.{hidden_projection.weight,predecessor_codebook,successor_codebook}`.

```
python3 -m sglang.launch_server \
  --model-path <qwen3.8-27b-autoround-int4> \
  --tp-size 2 --quantization auto-round --kv-cache-dtype fp8_e4m3 \
  --attention-backend flashinfer --context-length 262144 \
  --mem-fraction-static 0.82 --max-running-requests 1 \
  --speculative-algorithm DFLASH \
  --speculative-draft-model-path <DRAFTER> \
  --speculative-draft-model-quantization <compressed-tensors | unquant>
```

### Result

| drafter | `accept len` | `accept rate` | decode (code) |
|---|---:|---:|---:|
| syvai (compressed-tensors) | **1.03** (n=422) | **0.004** | ~38 tok/s |
| z-lab (BF16, `unquant`) | **3.71** (n=100) | 0.61 | ~171 tok/s |

With the quantized drafter, `accept len` never exceeded **1.12** across 422 decode batches. At
`block_size 8` a working drafter reads 2–5. Decode is **slower than running with no drafter at
all**, because it pays full verify cost for zero accepted tokens.

### Why this is easy to miss

1. The draft loads into the correct class and logs success:
   `Load weight end. type=DFlash2DraftModel, quant=compressed-tensors`.
2. Speculative decoding **rejects** bad drafts, so correctness is preserved by construction — a
   0%-acceptance drafter is slow, never wrong. Output stays fluent and correct.
3. Our functional test suite reported **"All checks passed"** on the broken configuration
   (completion, tool calls, streaming, reasoning split, a 2K-token coherence check, vision).

The only signal is the engine's own `accept len:` / `accept rate:` counters in the decode-batch log.

### What I ruled out

- **Not an architecture-mapping failure.** `DFlash2DraftModel` is first-class
  (`srt/models/dflash.py:1066`) and listed in
  `model_executor/model_runner_components/spec_aux_hidden_state.py`.
- **Not the target's aux-hidden-state capture.** #38589 reports that gap for `Qwen4Exp` and
  explicitly names `qwen3_5.py` — our target — as the *working* parent.
- **Not tensor naming.** Byte-for-byte identical between the two checkpoints.
- **Not the target `lm_head`.** Ours is dense BF16, satisfying the DFlash2 selector's dense path.

### Leading hypothesis (untested)

The `CandidateSelector` codebooks / `hidden_projection` are loaded through the compressed-tensors
path and end up wrong — well-formed but not the trained values — so the selector emits confident
garbage. That would match the symptom exactly: loads clean, runs, drafts nothing acceptable.

### Suggested mitigation regardless of root cause

Warn (or refuse) when a DFLASH draft resolves to a quantized checkpoint, as #36599 suggests for the
NextN path. A one-line boot warning would have saved a full benchmark cycle here.

### Related

- #36599 — same class (draft quant config mishandled), NextN/`modelopt_fp4`, fails loudly.
- #38589 — DFlash aux-hidden-state capture on `Qwen4Exp`.
- #36118 — runs `z-lab/Qwen3.8-27B-DFlash2` successfully (accepted span 4.49 @10K), consistent with
  the BF16 arm here.

### Environment

- SGLang `v0.5.19` (`lmsysorg/sglang:v0.5.19`), stock, no patches
- 2× RTX 3090 (sm_86), PCIe, no NVLink, TP=2, `--disable-custom-all-reduce`
- Downstream context: https://github.com/noonghunna/club-3090 (SGLang tier for Qwen3.8-27B)

⚠️ One SGLang-specific footgun found while narrowing this down, in case it is worth documenting
separately: `--speculative-draft-model-quantization` defaults to the **target's** quantization when
unset (`arg_groups/serving_hook.py:629`). Pointing at a BF16 drafter without passing `unquant`
therefore applies the target's `auto-round` to BF16 weights and reproduces the identical
0%-acceptance symptom from the other direction.

### PR DESCRIPTION
**Pull request by huihanglai**
## Motivation

Fixes #39087.

Pack-quantized DFlash2 draft checkpoints using `compressed-tensors` store the target-context projection `fc` as:

- `fc.weight_packed`
- `fc.weight_scale`
- `fc.weight_shape`

However, `DFlash2DraftModel` previously constructed `fc` as a plain `torch.nn.Linear`, which only exposes `fc.weight`.

As a result, the packed `fc` checkpoint tensors could not be resolved by `load_weights()` and were silently skipped. The `fc` layer therefore remained randomly initialized.

This does not prevent the model or server from starting, but causes the draft model to produce poor proposals. In the reported W4A16 DFlash2 case, speculative decoding acceptance collapses close to the bonus-token-only case.

## Modifications

This PR makes the DFlash2 target-context projection `fc` quantization-aware.

Main changes:

- Construct `fc` with SGLang's quantization-aware linear implementation when a quantization config is provided.
- Preserve the existing dense/BF16 loading path.
- Allow `compressed-tensors` pack-quantized checkpoints to correctly load:
  - `fc.weight_packed`
  - `fc.weight_scale`
  - `fc.weight_shape`
- Add explicit validation for incompatible dense `fc.weight` shapes instead of
  silently accepting an invalid checkpoint.
- Add CPU unit tests covering:
  - pack-quantized `fc` weight loading;
  - dense `fc.weight` loading;
  - invalid `fc.weight` shape handling.

The regression tests use small synthetic tensors and do not load real model checkpoints or launch a server.

## Accuracy Tests

The issue primarily affects speculative decoding acceptance rather than final target-model correctness. The target model still verifies generated tokens, but an incorrectly initialized draft `fc` causes most speculative tokens to be rejected.

The fixed W4A16 draft was compared with the BF16 reference draft under the same DFlash2 configuration:

- W4A16 draft: `syvai/Qwen3.8-27B-DFlash2-W4A16`
- BF16 reference draft: `incoai/Qwen3.8-27B-DFlash2`

| Metric | W4A16 Draft (After Fix) | incoai BF16 Draft |
| --- | ---: | ---: |
| Mean Accept Length | 3.23 | 3.25 |
| Median Accept Length | 3.36 | 3.30 |
| Accept Length Range | 2.27–4.10 | 2.12–4.05 |
| Mean Acceptance Rate | 0.32 | 0.31 |
| Median Acceptance Rate | 0.34 | 0.33 |
| Acceptance Rate Range | 0.18–0.44 | 0.16–0.44 |

After the fix, the W4A16 draft reaches essentially the same speculative acceptance behavior as the BF16 reference draft, indicating that the pack-quantized `fc` weights are now loaded correctly.

## Speed Tests and Profiling

Test hardware:
2 × NVIDIA RTX 3090
No NVLink

The fixed W4A16 draft was validated end-to-end with DFLASH speculative decoding enabled.
Runtime configuration:
Tensor parallel size:       2
Target model:               W4A16 AWQ / Marlin
Draft model:                W4A16 compressed-tensors
KV cache dtype:             FP8 E5M2
DFLASH block size:          8
Target verify CUDA graph:   enabled
Draft verify CUDA graph:    enabled
Max running requests:       2

After startup and warm-up, the quantized DFlash2 draft reached normal speculative acceptance. During sampled steady-state decode intervals for a code-generation workload, reported generation throughput ranged from approximately 86 to 260 tokens/s, depending strongly on the instantaneous acceptance length.
Example steady-state samples:
accept len: 3.45, accept rate: 0.35, gen throughput: 118.55 token/s
accept len: 5.10, accept rate: 0.59, gen throughput: 175.00 token/s
accept len: 6.88, accept rate: 0.84, gen throughput: 234.72 token/s
accept len: 7.72, accept rate: 0.96, gen throughput: 259.93 token/s

These values are sampled runtime statistics rather than a controlled before/after throughput benchmark. The primary performance regression criterion for this fix is the restoration of W4A16 speculative acceptance to the BF16 reference level.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations]  N/A — no user-facing behavior or API documentation changes. (https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #34732620479](https://github.com/sgl-project/sglang/actions/runs/34732620479)
Latest PR Test (Extra): :x: [Run #34732620367](https://github.com/sgl-project/sglang/actions/runs/34732620367)
Latest PR Test (AMD ROCm 10): :x: [Run #34732620429](https://github.com/sgl-project/sglang/actions/runs/34732620429)

### CODE REVIEW

**Comment by greatyingzi** [comment]
### Verified end-to-end on a quantized draft (2xRTX 3090, sglang 0.5.19)

We hit #39087 on a similar setup and worked around it in-tree; we then applied this patch on top of v0.5.19 and measured all three variants back-to-back in one window (same 5-prompt set, same sampling params, greedy for the first three).

Setup: 2xRTX 3090 (sm_86), sglang 0.5.19 native venv, tp=2, target `cyankiwi/Qwen3.8-27B-AWQ-INT4` (compressed-tensors), draft `syvai/Qwen3.8-27B-DFlash2-W4A16` (compressed-tensors), fp8 KV, `--mamba-radix-cache-strategy extra_buffer_lazy`, page-size 16, `--speculative-num-draft-tokens 8`.

| arm | code | sql | structured | prose (t=0.7) | chat (t=0.9) | agg | accept_length | accept_rate |
|---|---|---|---|---|---|---|---|---|
| v0.5.19 unpatched | 42.8 | 43.7 | 43.9 | 41.0 | 41.1 | 42.7 t/s | 1.10 | 1.4% |
| our in-tree workaround | 210.8 | 252.2 | 210.0 | 58.4 | 59.6 | 117.1 t/s | 1.65 | 9.3% |
| this PR applied | 227.7 | 256.9 | 214.3 | 58.0 | 65.5 | 115.3 t/s | 1.85 | 12.1% |

tok/s = `usage.completion_tokens` / wall time; `accept_*` read from `/metrics` (`sglang:spec_accept_length` / `spec_accept_rate`) right after the same prompt set in each arm.

Unpatched reproduces the collapse the issue describes (accept_length ~ bonus-token-only, 1.4% acceptance), and this PR restores the draft to the level of our in-tree fix — so the silently-dropped packed `fc` was the whole story on this checkpoint.

Two things worth settling before merge:

1. **`compute_dtype`**: this PR uses `self.fc.params_dtype`, while our in-tree version used `self.fc.weight.dtype`. For compressed-tensors packed weights those can differ (`weight` may be a packed int type). Both ran fine here, but we only exercise the DFlash2 path, not the Laguna/Nemotron aux-hidden path, so we cannot vouch for the choice — could you say why `params_dtype` is the correct one?
2. The now-**unconditional** `fc.weight` shape check did not reject our W4A16 checkpoint (it loaded and reached the numbers above). A case covering the previously-working dense path would be good insurance, if not already in the added unit tests.

Minor: the patch applies to v0.5.19 with offsets (hunks 2-7 within 17-79 lines, hunk 2 with fuzz 1), so a short rebase note may help whoever merges it after further `dflash.py` changes land.

PS - same class of bug elsewhere: several draft models still construct a plain `nn.Linear` while taking `quant_config` in the same file — `deepseek_nextn.py` (the site in #36599), `glm4_moe_nextn.py`, `glm4_moe_lite_nextn.py`, `glm_ocr_nextn.py`, `hunyuan_v3_nextn.py`, `mimo_v2_nextn.py` each have one such site. Might be worth generalizing the fix and the test.

cc @hnyls2002 @kpham-sgl - you reviewed the recent DFlash2 / DFLASH-mamba changes (#35371, #34763, #37818); this is the same code path.

**Comment by huihanglai** [comment]
For question 1
Since fc is now consistently constructed as ReplicatedLinear, I also switched the Laguna path to use the ReplicatedLinear interface consistently. input_size and **params_dtype** are members provided by LinearBase.
For question 2
Yes, this is already covered by the added CPU tests. test_dense_fc_still_loads constructs an unquantized DFlash2 draft with quant_config=None, loads a regular dense fc.weight, and verifies that it is loaded unchanged. test_fc_shape_mismatch_raises separately verifies that an incompatible dense fc.weight shape raises the expected ValueError. The pack-quantized path is independently covered by test_pack_quantized_fc_loads.

PS: I'm sorry, I don’t currently have enough resources to deploy and validate all models, such as deepseek and glm.
